src/NuGet.Core/NuGet.Packaging/Signing/Native/NativeCms.cs

```diff
@@ -48,15 +46,151 @@ private byte[] GetByteArrayAttribute(CMSG_GETPARAM_TYPE param, uint index)
             return data;
         }
 
-        internal static byte[] GetSignatureValueHash(HashAlgorithmName hashAlgorithmName, NativeCms nativeCms)
+        internal byte[] GetRepositoryCountersignatureSignatureValue()
         {
-            var signatureValue = nativeCms.GetEncryptedDigest();
+            var repositoryCountersignature = GetRepositoryCountersignature();
 
-            using (var signatureValueStream = new MemoryStream(signatureValue))
-            using (var hashAlgorithm = hashAlgorithmName.GetHashProvider())
+            if (repositoryCountersignature == null)
             {
-                return hashAlgorithm.ComputeHash(signatureValueStream, leaveStreamOpen: false);
+                return null;
             }
+
+            var countersignatureSignatureValue = new byte[repositoryCountersignature.Value.EncryptedHash.cbData];
+
+            Marshal.Copy(
+                repositoryCountersignature.Value.EncryptedHash.pbData,
+                countersignatureSignatureValue,
+                startIndex: 0,
+                length: countersignatureSignatureValue.Length);
+
+            return countersignatureSignatureValue;
+        }
+
+        private unsafe CMSG_SIGNER_INFO? GetRepositoryCountersignature()
+        {
+            const uint primarySignerInfoIndex = 0;
+            uint unsignedAttributeCount = 0;
+            var pointer = IntPtr.Zero;
+
+            NativeUtilities.ThrowIfFailed(NativeMethods.CryptMsgGetParam(
+                _handle,
+                CMSG_GETPARAM_TYPE.CMSG_SIGNER_UNAUTH_ATTR_PARAM,
+                primarySignerInfoIndex,
+                pointer,
+                ref unsignedAttributeCount));
+
+            if (unsignedAttributeCount == 0)
+            {
+                return null;
+            }
+
+            using (var retainer = new HeapBlockRetainer())
+            {
+                pointer = retainer.Alloc((int)unsignedAttributeCount);
+
+                NativeUtilities.ThrowIfFailed(NativeMethods.CryptMsgGetParam(
```

Comment:
Why are you calling this twice? One in line 75 and then this one.

Reply:
It's the classic "call once for size, call again for data" pattern in Win32 API's.  The `pointer` argument is [null](https://github.com/NuGet/NuGet.Client/pull/2017/files/542fea602342f68ece28fc387958a3c0a812ed1a#diff-9de9a2555ba61a56b5d822d0d90bf765R73) on the first call and [non-null](https://github.com/NuGet/NuGet.Client/pull/2017/files/542fea602342f68ece28fc387958a3c0a812ed1a#diff-9de9a2555ba61a56b5d822d0d90bf765R89) on the second call.  Here is the description for the `pvData` parameter from https://msdn.microsoft.com/en-us/library/windows/desktop/aa380227(v=vs.85).aspx:

    This parameter can be NULL to set the size of this information for memory allocation purposes.